core/src/test/java/kafka/server/SharePartitionManagerTest.java

```diff
@@ -1088,4 +1100,132 @@ public void testShareFetchContextResponseSize() {
+                k -> new SharePartition(groupId, tp1, 100, 5));
+        sharePartitionManager.partitionCacheMap.computeIfAbsent(sharePartitionManager.sharePartitionKey(groupId, tp2),
+                k -> new SharePartition(groupId, tp2, 100, 5));
+        sharePartitionManager.partitionCacheMap.computeIfAbsent(sharePartitionManager.sharePartitionKey(groupId, tp3),
+                k -> new SharePartition(groupId, tp3, 100, 5));
+
+        when(sp0.nextFetchOffset()).thenReturn((long) 1, (long) 15, (long) 6, (long) 30, (long) 25);
+        when(sp1.nextFetchOffset()).thenReturn((long) 4, (long) 1, (long) 18, (long) 5);
+        when(sp2.nextFetchOffset()).thenReturn((long) 10, (long) 25, (long) 26);
+        when(sp3.nextFetchOffset()).thenReturn((long) 20, (long) 15, (long) 23, (long) 16);
+
+        doAnswer(invocation -> {
+            assertEquals(1, sp0.nextFetchOffset());
+            assertEquals(4, sp1.nextFetchOffset());
+            assertEquals(10, sp2.nextFetchOffset());
+            assertEquals(20, sp3.nextFetchOffset());
+            return null;
+        }).doAnswer(invocation -> {
+            assertEquals(15, sp0.nextFetchOffset());
+            assertEquals(1, sp1.nextFetchOffset());
+            assertEquals(25, sp2.nextFetchOffset());
+            assertEquals(15, sp3.nextFetchOffset());
+            return null;
+        }).doAnswer(invocation -> {
+            assertEquals(6, sp0.nextFetchOffset());
+            assertEquals(18, sp1.nextFetchOffset());
+            assertEquals(26, sp2.nextFetchOffset());
+            assertEquals(23, sp3.nextFetchOffset());
+            return null;
+        }).doAnswer(invocation -> {
+            assertEquals(30, sp0.nextFetchOffset());
+            assertEquals(5, sp1.nextFetchOffset());
+            assertEquals(26, sp2.nextFetchOffset());
+            assertEquals(16, sp3.nextFetchOffset());
+            return null;
+        }).doAnswer(invocation -> {
+            assertEquals(25, sp0.nextFetchOffset());
+            assertEquals(5, sp1.nextFetchOffset());
+            assertEquals(26, sp2.nextFetchOffset());
+            assertEquals(16, sp3.nextFetchOffset());
+            return null;
+        }).when(replicaManager).fetchMessages(any(), any(), any(ReplicaQuota.class), any());
+
+        int threadCount = 100;
+        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
+
+        try {
+            for (int i = 0; i != threadCount; ++i) {
+                executorService.submit(() -> {
+                    sharePartitionManager.fetchMessages(groupId, memberId1.toString(), fetchParams, Arrays.asList(tp0, tp1, tp2, tp3));
+                });
+                // We are blocking the main thread at an interval of 10 threads so that the currently running executorService threads can complete.
+                if (i % 10 == 0)
+                    executorService.awaitTermination(2, TimeUnit.SECONDS);
```

Comment:
The mocked replicaManager `fetchMessages` completes instantaneously because there is no delay inside it. If we have to get delay present within the mocked `fetchMessages` function, the way we should implement it would be to have threads for individual fetch operations. If I use `thread.sleep()` right now, it will block everything, not just that thread.